docs/examples/ex02.py

```diff
@@ -69,7 +69,17 @@
 from skfem.models.poisson import unit_load
 import numpy as np
 
-m = MeshTri.init_symmetric().refined(3)
+m = (
+    MeshTri.init_symmetric()
+    .refined(3)
+    .with_boundaries(
+        {
+            "left": lambda x: x[0] == 0,
+            "right": lambda x: x[0] == 1,
+            "top": lambda x: x[1] == 1,
```

Comment:
I'm now using `np.isclose(...)` because I ran into some issues when the right hand side was non-integer float.

Reply:
Ouch. That's another reason for generally preferring to define boundaries at the CAD geometry stage rather than from a mesh.

I have had trouble using things like is-close when the extent of the geometry is physically very small (&micro;m, nm) but the coordinates have been measured in inappropriate units like SI metres. 

I try to only use float-equality when the float is being compared with a definite value that it may or may not have been assigned. I believe that that's reliable. (Whereas is-close would not be.) An exampke is the use of milestones in `pacopy.natural`.

In our functions that mesh segments, rectangles, and cuboids, aren't the terminal values of the coordinates preserved exactly?  ('Exactly' in the sense that floating-point equality with the same value is `True`, not exactly in the sense of the reals. )  So that `==` 0.0 or 1.0 should work for

https://github.com/kinnala/scikit-fem/blob/eb5a56562b19c3d525388f4a1298f78443ca35d1/docs/examples/ex02.py#L73

Reply:
I would argue that `==` in reals are either outright bugs or eggs that will turn into bugs when some other part of the code changes. In the case of `facets_satsifying` etc I think you're comparing to the midpoint of the facet. Even if `==` works for meshes generated by `skfem` it might fail meshes imported from other tools. Since you in general want people to copy the examples as starting points for their own problems, using `==` is probably putting bug-eggs in their code.

Reply:
would argue that == in reals are either outright bugs or eggs that will turn into bugs when some other part of the code changes

Reply:
> that == in reals are either outright bugs or eggs that will turn into bugs when some other part of the code changes

For my strenuous and ultimately successful argument against this widely accepted maxim (which should indeed be upheld almost everywhere), see nschloe/pacopy#6. Basically if x, a qnd b are float and one defines either x = a in one path or b in another then x == a (i.e. testing agzinst exactly the value possibly used to define the variable) is a meaningful test while a fuzzy comparison isn't since the nearness of b to a is unknown.

One way around this is what's in FreeFEM for their function that returns a meshed square.  I believe that such a function is an important special case and potentially worth treating specially. What they do is return the square with the four sides tagged; here we'd prepopulate the `.boundaries ` attribute with "left", "right", "floor", "ceiling". I think we might know the indices a priori without having to check the coordinates. However we often follow with `.refine` and i don't know whether rhat preserves boundaries (which would itself be a nice enhancement).

Reply:
refine (uniform) used to preserve tagged boundaries for MeshTri in the old version. Should be easy enough to copy paste that code. For other meshes requires some work.

Reply:
Your argument makes sense. And in the narrow case of your example, where I have assigned a or b to x, and later want to test for a or b, I agree it will work... however I think using floats for sentinel values like this is eventually going to introduce a bug.

In the case of the `skfem` examples though, it is not even obvious to me you fit this paradigm. When I create a mesh (for example a unit square mesh) in `gmsh`, read it into python with `meshio`, use `skfem.io.from_meshio` on it, and then ask for facets where `x[0]` is _exactly_ zero or one, I feel uneasy. I didn't explicitly assign zero to those points in the mesh.

But again, as you rightly point out, `isclose` is not a magic bullet. There's just no single right answer about how close you have to be. I think what you and I are both really doing in this thread is making the very strong case for labelling points during mesh creation instead of finding them with `facets_satsfying`.

Reply:
> I think what you and I are both really doing in this thread is making the very strong case for labelling points during mesh creation instead of finding them with facets_satsfying.

Yes, that's it. 100% Let's start working on that for the built-ins. I'll launch a discussion if you don't beat me to it.

> When I create a mesh (for example a unit square mesh) in gmsh, read it into python with meshio, use skfem.io.from_meshio on it, and then ask for facets where x[0] is exactly zero or one

Of this is really bad practice and to be discouraged, as you say. If one generates the mesh in Gmsh, the boundaries should be identified there. Considerable work has gone into having meshio and scikit-fem read and respect those boundaries from MSH 2.2 and 4.1 formats. 

I think the simole built-in canonical meshes do serve a role, avoiding recourse to external meshers, e.g., but not only, pedagogical examples, so let's get this right. Thanks!

Reply:
> In the case of the skfem examples though, it is not even obvious to me you fit this paradigm. When I create a mesh (for example a unit square mesh) in gmsh, read it into python with meshio, use skfem.io.from_meshio on it, and then ask for facets where x[0] is exactly zero or one, I feel uneasy. I didn't explicitly assign zero to those points in the mesh.

Note that this is not at all what is done in any of the skfem examples.  In exx 02, 03, 04 (`M`), 05, 11, 13, 25, 34, 42, which I believe are all of the examples that define boundaries using floating-point comparisons on coordinates, the meshes are not imported, they are generated using the internal functions and I believe that these functions do fit this paradigm (define `x = a`, then test `x == a`).  The examples that do import externally generated meshes, exx 04 (`m`), 17,  21, 24, 26 &ndash; 28, 35, 41, all already have boundaries or subdomains predefined in the externally generated mesh, as we're now recommending.

Reply:
> I'm now using `np.isclose(...)` because I ran into some issues when the right hand side was non-integer float.

@kinnala, could you share an MWE?  I can't reproduce this; e.g. `np.pi/2` works fine.

https://github.com/kinnala/scikit-fem/blob/76b27f8901eb710b8c6c2af4c5876d1fe880291b/docs/examples/ex13.py#L29-L35

Reply:
I can try to reproduce at some point. I was creating a mesh and then applied translate((0.1, 0)) and was trying to locate the right boundary after translate.

Reply:
Ah yes, I'd define boundaries before translating, as in 

https://github.com/kinnala/scikit-fem/blob/76b27f8901eb710b8c6c2af4c5876d1fe880291b/docs/examples/ex13.py#L29-L37

(for the reasons outlined above, viz. `x=a` then `x==a`, &c.)

Also on translation, I was wondering about why it was required in `ex04.M`

https://github.com/kinnala/scikit-fem/blob/57adf735347e4a05ea1541e883c91031041c1a73/docs/examples/ex04.py#L67-L71

instead of 

https://github.com/kinnala/scikit-fem/blob/76b27f8901eb710b8c6c2af4c5876d1fe880291b/docs/examples/ex04.py#L67-L70

(which I've taken the liberty of substituting in here).

Reply:
No reason whatsoever. :-)